server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1118,49 +1121,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                     }
                 ));
             }
+
+            // load latest repository data and run delete for all matching snapshots
+            private void resolveAndDeleteSnapshots(Collection<String> snapshotsOrPatterns, ActionListener<Void> listener,
+                                                   String repositoryName) {
+                assert snapshotsOrPatterns.isEmpty() == false;
+                threadPool.generic().execute(ActionRunnable.wrap(listener, l ->
+                        repositoriesService.repository(repositoryName).getRepositoryData(ActionListener.wrap(repositoryData ->
+                                deleteCompletedSnapshots(matchingSnapshotIds(repositoryData, snapshotsOrPatterns, repositoryName),
+                                        repositoryName, repositoryData.getGenId(), Priority.NORMAL, l), l::onFailure))));
+            }
         });
     }
 
+    private static List<SnapshotId> matchingSnapshotIds(RepositoryData repositoryData, Collection<String> snapshotsOrPatterns,
```

Comment:
I couldn't really bring my self to have this thing support `_all`, though I guess it still supports `*`. Both feel somewhat questionable/scary here, but I figured I'd at least not have `_all`.